Add vector search with embedding generation workload

vpehkone opened the pull request:
Signed-off-by: Vesa Pehkonen <vesa.pehkonen@intel.com>

### Description
Add vector search with embedding generation workload

### Issues Resolved
[#198](https://github.com/opensearch-project/opensearch-benchmark-workloads/issues/198) 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

akashsha1 commented:
@VijayanB / @wbeckler - PR for benchmarking vectorsearch with embedding i.e. neural search. Please help add necessary reviewers for this benchmark.

wbeckler commented:
What is the licensing of this content?

wbeckler commented on vectorsearch_embedding/README.md:
Hmmm, there might be an issue since OpenSearch is Apache License version 2.0 (see, e.g., https://huggingface.co/InstantX/InstantID/discussions/2)

akashsha1 replied:
I assume the issue is with the license of the MSMarco dataset? 
OpenSearch seems to already uses this dataset, and the msmarco sentence transformer model in other public documentation:
Neural Search tutorial: https://opensearch.org/docs/latest/search-plugins/neural-search-tutorial/
Blog: https://opensearch.org/blog/semantic-science-benchmarks/

This change is bringing the neural search pipeline into the benchmark repo.

Or do you suggest we use a different dataset with less restrictive licensing, and keep the other aspects of the workload?

vpehkone replied:
The dataset was changed to trec-covid.

navneet1v commented:
@vpehkone thanks for raising the PR. Can we move this benchmarks to folder named: semantic_search rather than vectorSearch_embedding

navneet1v commented on vectorsearch_embedding/workload.json:
```diff
@@ -0,0 +1,32 @@
+
+{
+  "version": 2,
+  "description": "MS MARCO is a collection of datasets focused on deep learning in search",
```
can we use trec-covid dataset instead of MSMarco? MS marco is a big dataset and might take a long time to complete.

vpehkone replied:
Dataset is not available on Trec-covid homepage https://ir.nist.gov/covidSubmit/index.html (dataset link does not work), also there is not anything about a license or if commercial use is allowed.

I found the trec-covid dataset here https://github.com/beir-cellar/beir and here https://huggingface.co/datasets/BeIR/trec-covid, but no license there.

Do you know where is it available with license?

navneet1v replied:
if the dataset is available on BeIR we can use the dataset. in general the default dataset we should set for this should be a one which good enough in size that it can be completed in limited time. My experience with MSMARCO is as it around 8M dataset, the whole benchmarks only ingestion can take time. Hence I was suggesting a smaller dataset.

akashsha1 replied:
@navneet1v - I see both MSMarco and Covid-trex available on BeIR: https://paperswithcode.com/dataset/beir

However, based on wbeckler and gkamat feedback on this PR: https://github.com/opensearch-project/opensearch-benchmark-workloads/pull/232#discussion_r1525618687, the license of the dataset cannot be 'non-commercial' .. Would you know how to verify covid-trex license information?

navneet1v replied:
@akashsha1 I don't have a way sorry.

akashsha1 replied:
@navneet1v  - thanks. I did some research, and confirmed that covid-trec is part of the COVID19 Open Research dataset. 

The covid-trec research paper[0] links to the Open Research Dataset[1], which is licensed under Apache 2.0. This should comply with OpenSearch license.
We should be good to proceed with this license for your PR @vpehkone 

cc: @wbeckler , @gkamat 

[0]: https://arxiv.org/pdf/2005.04474v1.pdf
[1]: https://github.com/allenai/cord19/tree/master

navneet1v replied:
This is awesome. :) thanks @akashsha1 for providing the details.

navneet1v commented on vectorsearch_embedding/workload.py:
```diff
@@ -0,0 +1,214 @@
+
+script_dir = os.path.dirname(os.path.realpath(__file__))
+
+class DeleteIngestPipeline(Runner):
```
is this operation not present in OSB? This is a pretty generic operation we should move this Operation of OSB.

vpehkone replied:
Moved this operation to OSB.

navneet1v commented on vectorsearch_embedding/workload.py:
```diff
@@ -0,0 +1,214 @@
+    def __repr__(self, *args, **kwargs):
+        return "delete-ingest-pipeline"
+
+class DeleteMlModel(Runner):
```
same as above.

vpehkone replied:
Moved this operation to OSB.

navneet1v commented on vectorsearch_embedding/workload.py:
```diff
@@ -0,0 +1,214 @@
+    def __repr__(self, *args, **kwargs):
+        return "delete-ml-model"
+
+class RegisterMlModel(Runner):
```
same as above

vpehkone replied:
Moved this operation to OSB.

navneet1v commented on vectorsearch_embedding/workload.py:
```diff
@@ -0,0 +1,214 @@
+        return "register-ml-model"
+
+
+class DeployMlModel(Runner):
```
same as above.

vpehkone replied:
Moved this operation to OSB.

vpehkone commented:
@navneet1v Addressed all comments: changed workload name to semantic_search, moved common code to OSB, and changed dataset to trec-covid. Can you please review?

IanHoang commented on semantic_search/test_procedures/default.json:
```diff
@@ -0,0 +1,126 @@
+        },
+        {
+          "operation": "semantic-search",
+          "warmup-iterations": {{warmup_iterations | default(100)}},
+          "iterations": {{iterations | default(100)}}
+          {%- if not target_throughput %}
+          ,"target-throughput": 10
+          {%- elif target_throughput is string and target_throughput.lower() == 'none' %}
+          {%- else %}
+          ,"target-throughput": {{ target_throughput | tojson }}
+          {%- endif %}
+          {%-if search_clients is defined and search_clients %}
+          ,"clients": {{ search_clients | tojson}}
+          {%- endif %}
```
We're aiming to simplify this for existing workloads (example [here](https://github.com/opensearch-project/opensearch-benchmark-workloads/issues/133)) and have implemented[ a different approach](https://github.com/opensearch-project/opensearch-benchmark-workloads/blob/aa95dce469bc256ec445254929d2523a8ccec5c3/big5/test_procedures/common/big5-schedule.json#L114-L118) as shown in the Big5 workloads. For better user experience, could you simplify this?

vpehkone replied:
Simplified. Now if we don't want to set the target throughput, we have to use empty string "" instead of "none".

IanHoang commented:
Since we have two semantic search workloads that are looking to be added, it might be beneficial to either rename both workloads to be more specific (such as `treccovid_semantic_search`) or have both contained under the same `semantic_search` folder, which might convolute the workload.

IanHoang commented on semantic_search/README.md:
````diff
@@ -0,0 +1,118 @@
+ ```
+
+Save it as `params.json` and provide it to OpenSearch Benchmark with `--workload-params="/path/to/params.json"`. The overrides for simple parameters could be specified in-place, for example `--workload-params=search_clients:2`.
+
````
Based on OSB workloads best practices, please add a sample output for what the user can expect to see when they run this workload.

vpehkone replied:
Added a sample output.

IanHoang commented on semantic_search/README.md:
```diff
@@ -0,0 +1,118 @@
+
+Save it as `params.json` and provide it to OpenSearch Benchmark with `--workload-params="/path/to/params.json"`. The overrides for simple parameters could be specified in-place, for example `--workload-params=search_clients:2`.
+
+### License
```
Please also include a section that briefly describes all the test procedures available to run in this workload.

vpehkone replied:
Added a description of procedures.

VijayanB commented:
> Since we have two semantic search workloads that are looking to be added, let's rename the workload to be more specific (such as `treccovid_semantic_search`).

@IanHoang If difference is just usage of datasets, do we need two different workloads? Why not merge and let individual procedures use their own corpus?

martin-gaievski commented on treccovid_semantic_search/test_procedures/default.json:
```diff
@@ -0,0 +1,112 @@
+          "operation": "deploy-ml-model"
+        },
+        {
+          "operation": "create-ingest-pipeline"
```
do you care about the ml-commons/model performance on doc ingestion? To get embeddings opensearch should do model inference

vpehkone replied:
Yes I care performance of indexing/embedding.

martin-gaievski commented on treccovid_semantic_search/test_procedures/default.json:
```diff
@@ -0,0 +1,112 @@
+          "operation": "wait-until-merges-finish"
+        },
+        {
+          "operation": "default",
```
this is simple match all query, curious why we need to collect throughput info for it in the context of neural search?

vpehkone replied:
The match all query was added because all other workloads include this.

martin-gaievski commented on treccovid_semantic_search/README.md:
```diff
@@ -0,0 +1,261 @@
+## Trec-Covid Semantic Search workload
+
+This workload uses OpenSearch pretrained model and ml-common-plugin to embed vectors. It is based on the neural search tutorial https://opensearch.org/docs/latest/search-plugins/neural-search-tutorial/ 
```
What are the main metrics you want to get with this workload? You are not checking correctness of the neural query results, is it mainly query latency?

vpehkone replied:
The main metrics are throughput and latency. No checking correctness, I will try to add this later.

vpehkone commented:
> > Since we have two semantic search workloads that are looking to be added, let's rename the workload to be more specific (such as `treccovid_semantic_search`).
> 
> @IanHoang If difference is just usage of datasets, do we need two different workloads? Why not merge and let individual procedures use their own corpus?

@VijayanB These workload are very different. Trec-covid semantic search generates embeddings and does vector search. Noaa semantic search does range, aggregate, term, etc... searches. It does not make sense to merge these.

martin-gaievski commented:
> > > Since we have two semantic search workloads that are looking to be added, let's rename the workload to be more specific (such as `treccovid_semantic_search`).
> > 
> > 
> > @IanHoang If difference is just usage of datasets, do we need two different workloads? Why not merge and let individual procedures use their own corpus?
> 
> @VijayanB These workload are very different. Trec-covid semantic search generates embeddings and does vector search. Noaa semantic search does range, aggregate, term, etc... searches. It does not make sense to merge these.

@vpehkone @VijayanB what would you say about rebasing that workload on a single dataset? We're not using this workload and trec-covid for checking correctness. We can use some field of text type to generate embeddings. Here is the example of doc from noaa:

```
{
  "TMIN": 8.3,
  "SNOW": "0",
  "WSF5": 11.2,
  "SNWD": "0",
  "PRCP": 0,
  "station": {
    "wmo_id": "72398",
    "state": "MARYLAND",
    "name": "SALISBURY WICOMICO RGNL AP",
    "location": {
      "lon": -75.5103,
      "lat": 38.3406
    },
    "elevation": 14.6,
    "country_code": "US",
    "id": "USW00093720",
    "country": "United",
    "state_code": "MD"
  },
  "WDF2": "20",
  "TMAX": 18.3,
  "AWND": 3.1,
  "TRANGE": {
    "lte": 18.3,
    "gte": 8.3
  },
  "date": "2015-04-15T00:00:00",
  "TAVG": 12.9,
  "WSF2": 8.1,
  "WDF5": "20"
}
```

we can use field `station.name` instead of `text` to generate embeddings.
Two workloads then can be merged into one as rest of operations are independent, maybe some operations can be reused. 
I'm not sure the other way around is possible as trec-covid mapping is simple and is lacking integer and keyword fields that are needed for hybrid query.

navneet1v commented:
@martin-gaievski what is the reason for merging the noaa dataset with Semantic Search? I think its better to keep semantic search as a separate use case and workload. Having a uber name as semantic search is really good. We can have more dataset in semantic search later. But for now having a simple Semantic Search dataset with trec-covid as dataset is pretty neat I would say.

martin-gaievski commented:
> @martin-gaievski what is the reason for merging the noaa dataset with Semantic Search? I think its better to keep semantic search as a separate use case and workload. Having a uber name as semantic search is really good. We can have more dataset in semantic search later. But for now having a simple Semantic Search dataset with trec-covid as dataset is pretty neat I would say.

I see few reasons to have them together:
- currently with noaa we cannot do hybrid queries with knn and neural sub-queries. such index mapping and custom operations are part of trec-covid workload. it would be an extra effort if we want to add it to noaa later
- trec-covid dataset is a bit small for measuring query latency. It has 170K doc, while noaa is 33M docs. this will give realistic latency results for neural/knn queries

navneet1v commented:
> * trec-covid dataset is a bit small for measuring query latency. It has 170K doc, while noaa is 33M docs. this will give realistic latency results for neural/knn queries

@martin-gaievski 33M docs ingestion via textembedding processor/MLpredict api will take a lot of time. Hence we should start with a small dataset. We can always add bigger dataset later.

Plus, if there are common part which can be used in Noaa and semantic search they should be moved to OSB code and should not live in workloads repo.

navneet1v commented:
> * currently with noaa we cannot do hybrid queries with knn and neural sub-queries. such index mapping and custom operations are part of trec-covid workload. it would be an extra effort if we want to add it to noaa later

For doing hybrid search with Noaa it will be hard with knn fields. The vector generation may take like days to complete. Hence for hybrid search with vector fields on bigger datasets we need to generate vectors once outside of OSB and directly ingest them rather than generating vectors on the fly.

martin-gaievski commented:
> > * currently with noaa we cannot do hybrid queries with knn and neural sub-queries. such index mapping and custom operations are part of trec-covid workload. it would be an extra effort if we want to add it to noaa later
> 
> For doing hybrid search with Noaa it will be hard with knn fields. The vector generation may take like days to complete. Hence for hybrid search with vector fields on bigger datasets we need to generate vectors once outside of OSB and directly ingest them rather than generating vectors on the fly.

That long running operation can be a benchmark by itself, but if it takes days and we're focusing only of performance of search then I'm fine with smaller dataset like trec-covid

vpehkone commented:
@martin-gaievski  , @navneet1v  - based on the discussion, it seems it is better to keep the two workloads separate. That is how the trec-covid PR is currently designed. If there are no other issues on the PR, could you help merge? thanks.

IanHoang commented:
@vpehkone does the data corpora need to be added to a cloud repository (similar to [this workload.json](https://github.com/opensearch-project/opensearch-benchmark-workloads/blob/main/big5/workload.json#L18)) If so, we might need to add a files.txt, [similar to other workloads](https://github.com/opensearch-project/opensearch-benchmark-workloads/blob/main/nyc_taxis/files.txt).

vpehkone commented:
> @vpehkone does the data corpora need to be added to a cloud repository (similar to [this workload.json](https://github.com/opensearch-project/opensearch-benchmark-workloads/blob/main/big5/workload.json#L18)) If so, we might need to add a files.txt, [similar to other workloads](https://github.com/opensearch-project/opensearch-benchmark-workloads/blob/main/nyc_taxis/files.txt).

@IanHoang Yes, I will add files.txt. Can you let me know how add the data corpora to the cloud repository? Then I can update the corpus url at the same time.

VijayanB commented on treccovid_semantic_search/workload.json:
```diff
@@ -0,0 +1,32 @@
+  "corpora": [
+    {
+      "name": "treccovid",
+      "base-url": "https://vesa-oswl.s3.us-west-2.amazonaws.com/treccovid",
```
Who is maintaining this service?

vpehkone replied:
Its my aws account. I can copy the corpus to the opensearch official cloud, if I can get instruction how to do it.

martin-gaievski replied:
one option for hosting the dataset is to add it as a release artifact to one of opensearch repos (existing or new). This way it will be hosted at github site and will be available to public with direct link

VijayanB commented on treccovid_semantic_search/workload.py:
```diff
@@ -0,0 +1,75 @@
+                params['body']['query']['neural']['passage_embedding']['query_text'] = query_text
+        return params
+
+def register(registry):
+    registry.register_param_source("semantic-search-source", QueryParamSource)
+    registry.register_param_source("create-ingest-pipeline", ingest_pipeline_param_source)
```
@IanHoang Can this be added directly to opensearch-benchmark-repository?

IanHoang replied:
@VijayanB Sorry for the lates response. Yes, @vpehkone this can be technically contributed directly into [the OpenSearch Benchmark repository ](https://github.com/opensearch-project/opensearch-benchmark)

vpehkone replied:
@IanHoang I don't know how it could be possible to add a registration of parameter source function/class to OSB. OSB cannot know where to look at it or the name of the source function/class? Please let me know if you have an idea how it works, and I will try to implement it.

IanHoang replied:
@vpehkone We plan to add some documentation to the official OSB documentation regarding this. In the interest of time, we can have this merged in

VijayanB commented on treccovid_semantic_search/README.md:
```diff
@@ -0,0 +1,261 @@
+* `k` (default: 10) Number of nearest neighbors are returned.
+* `warmup_iterations` Number of Warmup iteration of each search client executes.
+* `iterations`  Number of test iterations of each search client executes.
+* `variable_queries` (default: 0) Number of variable queries will be used for the semantic search task, 0 means fixed query and max value is 20,000.
```
may be num_variable_queries?

VijayanB commented on treccovid_semantic_search/index.json:
```diff
@@ -0,0 +1,30 @@
+        "type": "knn_vector",
+        "dimension": {{dimensions | default(768)}},
+        "method": {
+          "engine": "{{engine | default('lucene')}}",
```
why lucene is default? why not set only engine if it is defined ?

vpehkone replied:
Done, no default engine.

VijayanB commented on treccovid_semantic_search/index.json:
```diff
@@ -0,0 +1,30 @@
+{
+  "settings": {
+    "index.number_of_shards": {{number_of_shards | default(1)}},
```
lets set this value only if it is defined

vpehkone replied:
Done, no default num of shards.

VijayanB commented on treccovid_semantic_search/index.json:
```diff
@@ -0,0 +1,30 @@
+{
+  "settings": {
+    "index.number_of_shards": {{number_of_shards | default(1)}},
+    "index.number_of_replicas": {{number_of_replicas | default(0)}},
```
same as above

vpehkone replied:
Done, no default replicas.

VijayanB commented on treccovid_semantic_search/test_procedures/default.json:
```diff
@@ -0,0 +1,112 @@
+          "operation": "refresh"
+        },
+        {
+          "operation": "wait-until-merges-finish"
```
why do we need this operation?

vpehkone replied:
We wait that merges are finished before searching. The same is in other workloads, e.g. nyc_taxis https://github.com/opensearch-project/opensearch-benchmark-workloads/blob/main/nyc_taxis/test_procedures/default.json

VijayanB commented on treccovid_semantic_search/workload.py:
```diff
@@ -0,0 +1,75 @@
+def ingest_pipeline_param_source(workload, params, **kwargs):
+    model_id = params['body']['processors'][0]['text_embedding']['model_id']
+    if not model_id:
+        with open('model_id.json') as f:
```
I don't see model_id.json in your commit. How are you importing this file into your workload?

vpehkone replied:
RegisterMlModel() returns a model_id and saves it to json file.

VijayanB commented on treccovid_semantic_search/workload.py:
```diff
@@ -0,0 +1,75 @@
+        self._params['type'] = type
+        self._params['variable-queries'] = params.get("variable-queries", 0)
+        self.infinite = True
+
+        if self._params['variable-queries'] > 0:
+            with open(script_dir + os.sep + 'workload_queries.json', 'r') as f:
+                d = json.loads(f.read())
+                source_file = d['source-file']
+                base_url = d['base-url']
+                compressed_bytes = d['compressed-bytes']
+                uncompressed_bytes = d['uncompressed-bytes']
+                compressed_path = script_dir + os.sep + source_file
+                uncompressed_path = script_dir + os.sep + Path(source_file).stem
+            if not os.path.exists(compressed_path):
+                downloader = Downloader(False, False)
+                downloader.download(base_url, None, compressed_path, compressed_bytes)
+            if not os.path.exists(uncompressed_path):
+                decompressor = Decompressor()
+                decompressor.decompress(compressed_path, uncompressed_path, uncompressed_bytes)
```
can this be moved as corpora and get rid of this logic?

vpehkone replied:
It cannot, because all data in corpora is bulk uploaded to index and we don't want to index queries. Also if we run only the search task, we would not get queries.

martin-gaievski commented on treccovid_semantic_search/workload.json:
```diff
@@ -0,0 +1,32 @@
+  ],
+  "corpora": [
+    {
+      "name": "treccovid",
```
Is the default trec-covid dataset big enough to give reliable information about latency? It has about 170K documents. We are going to host the corpora anyways, can we pre-process it and increase the number of documents by creating multiple copies of the same doc? Like 6 copies, this should give us about 1M.

vpehkone replied:
1M documents works for me and it would also solve the dataset hosting issue, but the queries should be hosted in the same place.

martin-gaievski replied:
I think more on this option, probably it's better to make it an enhancement to this workload or even a separate workload. I think we are good for this PR with what we have

IanHoang commented:
> > @vpehkone does the data corpora need to be added to a cloud repository (similar to [this workload.json](https://github.com/opensearch-project/opensearch-benchmark-workloads/blob/main/big5/workload.json#L18)) If so, we might need to add a files.txt, [similar to other workloads](https://github.com/opensearch-project/opensearch-benchmark-workloads/blob/main/nyc_taxis/files.txt).
> 
> @IanHoang Yes, I added files.txt. Can you let me know how add the data corpora to the cloud repository? Then I can update the corpus url to workload.json.

We'll need to download the corpora manually and add it to our cloud repository. I can coordinate this with you offline in the slack community.

IanHoang commented:
@vpehkone Please add this to the workload.json as the base-url: https://dbyiw3u3rf9yr.cloudfront.net/corpora/treccovid

IanHoang commented on treccovid_semantic_search/workload.json:
```diff
@@ -0,0 +1,32 @@
+  "corpora": [
+    {
+      "name": "treccovid",
+      "base-url": "https://vesa-oswl.s3.us-west-2.amazonaws.com/treccovid",
```
@vpehkone This can now be updated to `https://dbyiw3u3rf9yr.cloudfront.net/corpora/treccovid`. Please confirm if document count is also correct. Is `129192` from the uncompressed version of documents.json? 
```
$ wc -l documents.json.bz2
253809 documents.json.bz2
```

vpehkone replied:
Updated the new URL for documents and queries. Yes, the document count is right.

navneet1v commented:
excited to see this workload getting merged. :) thanks @vpehkone making this happen and @IanHoang for reviewing the code.
